bottlerocket-os/bottlerocket
# k8s settings: kube-reserved, eviction-hard, cpu-manager-policy, allow-unsafe-sysctls

**My fork draft PR**: https://github.com/gthao313/bottlerocket/pull/1


### **Description of changes:**

- Add new settings `eviction-hard`, `kube-reserved`, `allowed-unsafe-sysctls` to k8s. User can specify this setting in userdata.
- Enable `cpu-manager-policy` to static

- Add migration for `eviction-hard`, `kube-reserved`, `allowed-unsafe-sysctls` 


**example**: 

```
[settings.kubernetes.eviction-hard]
"memory.available" = "10Gi"
"nodefs.inodesFree" = “15%”
[settings.kubernetes.kube-reserved]
cpu = "70Mi"
memory = "1Gi"
ephemeral-storage = "1Gi"
[settings.kubernetes]
allowed-unsafe-sysctls = ["net.core.somaxconn", ...]
```

### **Testing done:**

- Build aws-eks-* images and launched instance.
- Migration test: tested the migration via upgrade and downgrade testing
- Unit tests for new model types `EvictionHardKey`, `KubernetesThresholdValue`, `ReservedResourcesKey`, `KubernetesQuantityValue`, and `CpuManagerPolicy` (all passes)

### Eviction-hard test
***Test step1***:
Go to control container run `apiclient -u /settings` to check if expected setting is there.

***Test step2***:
ssh to admin contianer and look `etc/kubernetes/kubelet/config` to check if expected config is there.
```
evictionHard:
  memory.available: 15%
```
***Test step3*** - test effected/behavior:
set up eviction threshold to a high value for triggering eviction.
```
[evictionHard.memory]
"memory.available" = "99.99%" (high percentage to trigger eviction)
```
a) Go to EKS to check node condition 

MemoryPressure | True | kubelet has insufficient memory available
-- | -- | --

b) In admin contianer, check kubelet log by `journalctl -u kubelet`
```
eviction_manager.go:335] eviction manager: attempting to reclaim memory
eviction_manager.go:346] eviction manager: must evict pod(s) to reclaim memory
```
### kube-reservd test
***Test step1***:
Go to control container run apiclient -u /settings to check if expected setting is there.

***Test step2***:
_NodeAllocatable = NodeCapacity - Kube-reserved - system-reserved - eviction-threshold_
**cpu**
[settings.kubernetes.kube-reserved.cpu]
cpu = 1
Allocatable cpu = 2 - 1 = 1
	
```
EKS Resource allocation:
Name Capacity Allocatable
CPU		2		1
```

**memory**
[settings.kubernetes.kube-reserved.Memory]
memory = 1Gi

```
EKS Resource allocation:
Name    Capacity    Allocatable
Memory	7844780Ki	6693804Ki
```

**ephemeral-storage**
[settings.kubernetes.kube-reserved.ephemeral-storage]
ephemeral-storage = 1Gi

```
Capacity:
  attachable-volumes-aws-ebs:  25
  cpu:                         2
  ephemeral-storage:           20624592Ki
  hugepages-1Gi:               0
  hugepages-2Mi:               0
  memory:                      7930796Ki
  pods:                        29
Allocatable:
  attachable-volumes-aws-ebs:  25
  cpu:                         1
  ephemeral-storage:           19007623956
  hugepages-1Gi:               0
  hugepages-2Mi:               0
  memory:                      7828396Ki
  pods:                        29˜
```
### cpu-manager-policy test
***Test step1***:
Go to host find `cpu_manager_state`
```
bash-5.0# cat /var/lib/kubelet/cpu_manager_state
{"policyName":"static","defaultCpuSet":"0-1","checksum":3945352861}bash-5.0#
```

```
journalctl -u kubelet
cpu_manager.go:173] [cpumanager] starting with static policy
```
### allowed-unsafe-sysctls test
***Test step1***:
Go to control container run apiclient -u /settings to check if expected setting is there.

***Test step2***:
Go to kubelet-config to check if desired result is there.
```
cat eks/kubernetes/kubelet/config

allowed-unsafe-sysctls = ["net.core.somaxconn", ]
```



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
